src/codelabs/null-safety.md

```diff
@@ -0,0 +1,336 @@
+can't possibly contain null values are treated like non-nullable variables.
+This behavior is called
+[type promotion](/null-safety/understanding-null-safety#type-promotion-on-null-checks).
+
+### Example: Definite assignment
+
+Dart's type system can track where variables are assigned and
+where their values are read, and to verify that non-nullable fields are given
+values before any code tries to read from them. This process is called
+[definite assignment](/null-safety/understanding-null-safety#definite-assignment-analysis).
+
+Try uncommenting the if-else statement in the code below, and
+watch the analyzer errors disappear:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/definite_assignment.dart" replace="/if/\/\/if/g; /\ \ text\ =/\/\/  text =/g; /\ \ \}/  \/\/}/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  String text;
+
+  //if (DateTime.now().hour < 12) {
+  //  text = "It's morning! Let's make aloo paratha!";
+  //} else {
+  //  text = "It's afternoon! Let's make biryani!";
+  //}
+
+  print(text);
+  print(text.length);
+}
+```
+
+### Exercise: Null checking
+
+In the example below, add an if statement to the beginning of `getLength` that
+returns zero if `str` is null:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/type_promotion.dart" replace="/.*if\ \(.*\n.*\n.*//g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+int getLength(String? str) {
+  // Add null check here
+
+  return str.length;
+}
+
+void main() {
+  print(getLength('This is a string!'));
+}
+```
+
+### Exercise: Promotion with exceptions
+
+Promotion works with exceptions as well as return statements. Try a null check
+that throws an `Exception` instead of returning zero.
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/promotion_exceptions.dart" replace="/.*if\ \(.*\n.*\n.*//g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+int getLength(String? str) {
+  // Try throwing an exception here if `str` is null.
+
+  return str.length;
+}
+
+void main() {
+  print(getLength(null));
+}
+```
+
+## Late keyword
+
+Sometimes fields in a class *should* be non-nullable, but those fields can't be
+assigned a value immediately. For cases like that, use the
+[`late` keyword]( /null-safety/understanding-null-safety#late-variables).
+
+`late` is how you tell Dart that:
+
+-   You aren't going to assign that field a value right away.
+-   But you *are* going to assign it a value later.
+-   And you'll make sure it's assigned a value *before* it's accessed.
+
+If you declare a field `late` and the field is read before it is assigned a
+value, a `LateInitializationError` is thrown to tell you what went wrong.
+
+### Example: The `late` keyword
+
+Try using the `late` keyword to correct the following code. For a little extra
+fun afterward, try commenting out the line that sets `description`!
+
```

Comment:
Have you noticed that when you click **Reset**, analysis isn't run again? I've filed https://github.com/dart-lang/dart-pad/issues/1785 about this.